### PR TITLE
py-h5py: offline installation and deps fix

### PR DESCRIPTION
**Pull request by bryanherman**
Since h5py v3, offline installations have not worked. This is because of how h5py developers changed the structure of setup.py (https://github.com/h5py/h5py/blob/3.2.1/setup.py). If you look carefully, the SETUP_REQUIRES is using very strict version dependencies for numpy that are allowed to be different than the RUN_REQUIRES. This causes a problem in Spack because it expects them to be the same version. So there are two options: use the strict versions for setup requires for runtime deps as well, or patch h5py's setup.py to loosen the == restriction for setup_requires. I chose to do the latter because h5py built fine. I will work with h5py developers separately to understand why they went that route, but I wanted to get something into Spack that works for offline installations. If one looks at the Spack build output, if the following warning appears, you know that offline installation wont work:

```
WARNING: The pip package is not available, falling back to EasyInstall for handling setup_requires/test_requires; this is deprecated and will be removed in a future version.
```

In addition to supporting offline installation, additional fixes were made:

- the python dependency listed in py-cached-property did not have a ^
- Cython version deps didn't match setup.py
- NumPy version deps didn't match setup.py
- MPI4Py version deps didn't match setup.py

### CODE REVIEW

**Comment by adamjstewart** [comment]
> So there are two options: use the strict versions for setup requires for runtime deps as well, or patch h5py's setup.py to loosen the == restriction for setup_requires.

I actually prefer the former in the short-term since patches may need to be updated frequently if the code keeps changing. But in the long-term, if you can figure out why the developers chose to do this and convince them to undo this, then your patch is the better solution.

**Review by adamjstewart — APPROVED** [review]
This looks great. Would you have any interest in being listed as the official maintainer for this recipe? If so, just add:
```python
maintainers = ['bryanherman']
```
You don't have to be a Spack/h5py expert, it just gives us someone to ping to review PRs or build issues for the package.

**Comment by bryanherman** [comment]
> This looks great. Would you have any interest in being listed as the official maintainer for this recipe? If so, just add:
> 
> ```python
> maintainers = ['bryanherman']
> ```
> 
> You don't have to be a Spack/h5py expert, it just gives us someone to ping to review PRs or build issues for the package.

I think I will do this, I need to get some local approvals as our organization is newer to this type of collaboration. I'll make a separate PR once I'm good to go.

**Comment by bryanherman** [comment]
> > So there are two options: use the strict versions for setup requires for runtime deps as well, or patch h5py's setup.py to loosen the == restriction for setup_requires.
> 
> I actually prefer the former in the short-term since patches may need to be updated frequently if the code keeps changing. But in the long-term, if you can figure out why the developers chose to do this and convince them to undo this, then your patch is the better solution.

You are right that patching is more difficult to maintain, however I think if we restrict folks to a specific numpy version in h5py there will unnecessarily be less flexibility in the higher level applications. So I'd like to keep the patch for now, work with h5py folks and then contribute the conclusion of that effort later.

**Comment by takluyver** [comment]
Specifying particular versions of NumPy & mpi4py as build dependencies is kind of a workaround for the limitations of Python packaging: when a tool like `pip` builds h5py from source, we want it to build with the oldest compatible NumPy, which means the resulting package will work alongside that or any newer NumPy.

For something like Spack, which manages and records things like this separately, the easiest way to avoid these limitations is to set the environment variable `H5PY_SETUP_REQUIRES=0`. That removes all the `setup_requires`, so it's trusting that the build environment already contains the necessary packages.

**Pull request merged**
